include/envoy/network/connection.h

```diff
@@ -116,6 +116,13 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
   virtual const std::string& remoteAddress() PURE;
 
   /**
+   * @return The address the remote client is trying to connect to.
+   * It can be different from the proxy address if the downstream connection
+   * has been redirected or the proxy is operating in transparent mode.
+   */
+  virtual const std::string destinationAddress() PURE;
```

Comment:
In thinking about this more, this is still confusing as this is a base class for both client and server connections. Here is what I would which shouldn't be too hard: rename this function to localAddress(). For server connections, local address would be the incoming address, and for client connections local address would be the origin address. I think this makes sense.

I will comment on some implementation stuff below.